File: src/op/sigmoid.cc
#include "op/sigmoid.h"

#include "utils/utils.h"
#include "tensor/tensor.cuh"
#include "backend/cuda/sigmoid.cuh"

namespace TinyTorch {

Tensor sigmoid_forward(const Tensor &input) {
    ASSERT(input.device.is_cuda());
    Tensor output(input.shape, input.device);
    if (input.device.is_cuda()) {
        Backend::CUDA::sigmoid_forward(input.data, output.data, input.numel());
    } else {}
    return output;
}

Tensor sigmoid_backward(const Tensor &output, const Tensor &grad_output) {
    ASSERT(output.device.is_cuda() && grad_output.device.is_cuda());
    ASSERT(output.shape == grad_output.shape);
    Tensor grad_input(output.shape, output.device);
    if (output.device.is_cuda()) {
        Backend::CUDA::sigmoid_backward(grad_output.data, output.data, grad_input.data, output.numel());
    } else {}
    return grad_input;
}

}

File: unittest/sigmoid.py
import tinytorch

import torch
from torch import nn
from torch.nn import functional as F

import numpy as np

SHAPE = [32, 3, 28, 28]

device = None

def test_sigmoid():
    input = torch.randn(SHAPE, device=device, requires_grad=True)
    grad_output = torch.randn(SHAPE, device=device, requires_grad=True)
    output_gt = F.sigmoid(input)
    output_gt.backward(grad_output)

    output_gt = tinytorch.TensorBase(output_gt.detach().numpy())
    grad_input_gt = tinytorch.TensorBase(input.grad.detach().numpy())
    input = tinytorch.TensorBase(input.detach().numpy())
    grad_output = tinytorch.TensorBase(grad_output.detach().numpy())

    output_res = tinytorch.op.sigmoid_forward(input)
    grad_input_res = tinytorch.op.sigmoid_backward(output_res, grad_output)

    assert output_res == output_gt
    assert grad_input_res == grad_input_gt

def main():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print("Torch Device:", device)
    tinytorch.Device.set_default_device(tinytorch.Device.cuda())
    print("TinyTorch Default Device:", "cpu" if tinytorch.Device.get_default_device().is_cpu() else "cuda")
    test_sigmoid()

if __name__ == "__main__":
    main()
